<!DOCTYPE html>
<html>
	<head>
		<meta charset="utf-8" />
		<meta name="viewport" content="width=device-width, initial-scale=1" />
		<title>Mechnetor</title>
		<link
			rel="shortcut icon"
			type="image/x-icon"
			href="static/images/mechnetor.ico"
		/>
		<link
			rel="stylesheet"
			href="https://fonts.googleapis.com/css?family=Staatliches|Lato|Audiowide|Source+Sans+Pro:400,600"
		/>
		<link
			rel="stylesheet"
			type="text/css"
			href="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/css/bootstrap.min.css"
			integrity="sha384-MCw98/SFnGE8fJT3GXwEOngsV7Zt27NXFoaoApmYm81iuXoPkFOJwJ8ERdknLPMO"
			crossorigin="anonymous"
		/>
		<link
			rel="stylesheet"
			type="text/css"
			href="https://use.fontawesome.com/releases/v5.2.0/css/all.css"
			integrity="sha384-hWVjflwFxL6sNzntih27bfxkr27PmbbK/iSvJ+a4+0owXq79v+lsFkW54bOGbiDQ"
			crossorigin="anonymous"
		/>
		<link
			rel="stylesheet"
			type="text/css"
			href="{{ url_for('static',filename='css/common.css') }}"
		/>
		<script
			type="text/javascript"
			src="{{ url_for('static',filename='js/jquery-3.3.1.js') }}"
		></script>

		<script
			type="text/javascript"
			src="https://cdnjs.cloudflare.com/ajax/libs/jquery-validate/1.17.0/jquery.validate.min.js"
		></script>
		<script
			type="text/javascript"
			src="https://cdnjs.cloudflare.com/ajax/libs/jquery-validate/1.17.0/additional-methods.js"
		></script>

		<style>
			.main-header {
				margin-top: 10px;
			}
			.sub-header {
				border-bottom: 1px solid #000;
				padding-bottom: 2px;
				font-size: large;
				font-weight: bold;
				margin-bottom: 5px;
			}
			.main-container {
				max-width: 60%;
				margin: auto;
				background-color: #fff;
				padding: 10px;
				border: 1px solid #ccc;
			}
			table {
				width: 400px;
				margin-bottom: 5px;
			}
			tr {
				border-bottom: 1px solid #ddd;
				padding-right: 20px;
			}
		</style>
	</head>

	<body>
		<div class="grid-container">
			{% include "header.html" %}

			<div class="content">
				<div class="main-container">
					<div class="main-header">
						<h3>
							<i class="fas fa-question-circle"></i
							><b> Help and documentation</b>
						</h3>
					</div>

					<div class="sub-header">What is Mechnetor?</div>
					<p>
						The current abundance of genetic variant data requires new tools to
						develop mechanistic hypotheses about biological processes and
						diseases. Fortunately, growing datasets of protein interactions,
						mechanism and structure can provide help understand or predict
						biological mechanism. However, a systematic integration of these
						datasets is first needed. For this purpose, we developed
						<strong>Mechnetor</strong>. <br /><br />
						For sets of proteins or interactions, the tool integrates
						mechanistic data from diverse sources (below) and constructs an
						interactive network for intuitive visualization of interaction
						mechanisms. Proteins are represented as linear arrangements of
						domains, motifs and other functional elements, which permits the
						display of interactions between the relevant/known protein regions.
						The result is a finer resolution interaction network and more
						immediate clues as to points of intervention. The view can be
						customized and manipulated, and exported as an image or vector
						graphics (SVG) at any time, providing beautiful figures suitable for
						publication.
					</p>

					<div class="sub-header">Input</div>
					<p>
						<strong>Mechnetor</strong> requires protein pairs so that it can
						look for interactions between them. Users can provide their input in
						different ways but always using <b>gene names</b>,
						<b>UniProt accessions</b> or <b>UniProt identifiers</b>.
						<br /><br />
						You can input a <b>list of proteins</b> (one protein per line) and
						it will look for interactions between every possible protein pair.
						If you want to restric it to particular pairs, then enter a
						<b>list of protein pairs</b> (two proteins in the same line
						space-separated). <br />
						Using the slider below, you can automatically add
						<b>a number of known interactors</b> per each protein. These are
						always retrieved in order of the amount of evidence according to the
						PPI database. This option makes possible simple queries of a
						<b>single protein</b> and few of its most common interactors.
						<br /><br />
						You can input any number of
						<b>protein mutations or modifications</b> (always one per line),
						following the specified format. Mutations will be mapped to their
						corresponding positions within the protein but they are not
						neccessary to run a job. At the same time, any protein in the
						mutation input box will be added to the network, thus it's not
						neccessary to enter it on both boxes.
					</p>

					<!-- <ul>
						<li>Input a list of proteins</li>
						<img src="{{ url_for('static',filename='images/inputs1.png') }}"
						width="150" height="150"
						style="margin:auto;"
						>
						<li>Tea</li>
						<li>Milk</li>
					  </ul>  -->

					<div class="sub-header">Output</div>
					<p>
						<strong>Mechnetor</strong> creates a mechanistic network with the
						proteins and variants you provided, with proteins and their different functional
						elements arranged linearly, and variants displayed on them.
						This network is highly interactive and the user is encouraged to manually explore 
						their results using the mouse controls and several toggle buttons and sliders (see 
						the Quick Guide). 
						<br><br>
						The current network view can be exported as an image (JPG, PNG) or vector
						graphics (SVG) at any time.
						<br><br>
						The different interaction evidence displayed in the network (domain-domain interactions, 
						domain-motif interactions, interacting regions from annotation, predicted interaction interfaces)
						is also contained in a browsable list, that can also be downloaded in different formats. 
						<br><br>
						The results page can be bookmarked and accessed later. Results will be kept for 
						at least 6 months
					</p>

					<br /><br />

					<div class="main-header">
						<h3><i class="far fa-chart-bar"></i><b> Data</b></h3>
					</div>

					<p>This are all the data sources currently used by Mechnetor:</p>
					<table>
						<tr style="border-bottom: 2px" solid #ddd>
							<th>Database</th>
							<th>Version/Release</th>
						</tr>
						<tr>
							<td>UniProt</td>
							<td>2020_05</td>
						</tr>
						<tr>
							<td>Pfam</td>
							<td>33.1</td>
						</tr>
						<tr>
							<td>ELM</td>
							<td>March 2020</td>
						</tr>
						<tr>
							<td>3did</td>
							<td>2020-01</td>
						</tr>
						<tr>
							<td>BioGRID</td>
							<td>4.2.191</td>
						</tr>
						<tr>
							<td>PhosphoSitePlus</td>
							<td>March 2019</td>
						</tr>
						<tr>
							<td>COSMIC</td>
							<td>v87</td>
						</tr>
					</table>
					<br />
					<p>
						It also runs
						<a
							href="https://http://www.russelllab.org/cgi-bin/tools/interprets.pl"
							>InterPreTS</a
						>, our tool to predict interactions through tertiary structure.
					</p>
					<p>Currently, it supports human and 6 other model organisms.</p>
				</div>
			</div>

			{% include "footer.html" %}
		</div>
	</body>
</html>
